[wip][clustering] visualization dashboard

mivanit opened the pull request:
## Description
wip, marked as "ready for review" for getting CI tests to run

see #198 
<!--- Describe your changes in detail -->

## Related Issue
<!--- Use the keywords 'Close', 'Closes', 'Fix', or 'Fixes' followed by the issue number(s) to close the related issue(s) -->
#133 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->

## Does this PR introduce a breaking change?
<!--- If this PR introduces a breaking change, please describe the impact and migration path for existing applications below. -->

mivanit committed:
comments explaining port allocation in tests

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

danbraunai-goodfire committed:
Merge branch 'main' into feature/clustering

mivanit committed:
add distributed marker, rull all distributed tests on same worker

mivanit committed:
add distributed marker, rull all distributed tests on same worker

mivanit committed:
Revert "add distributed marker, rull all distributed tests on same worker"

This reverts commit 3f55ffa932dfa2c4d0e0c2bcb69ea6615528c049.

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
Merge branch 'main' into refactor/clustering-prereqs

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
add "num_nonsingleton_groups" metric to spd-cluster

mivanit committed:
uv sync

mivanit committed:
wip jaccard

mivanit committed:
make format

mivanit committed:
refactor: use general_utils methods for getting device everywhere

mivanit committed:
Merge branch 'main' into refactor/clustering-prereqs

mivanit committed:
Merge branch 'main' into feature/clustering

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
wip jaccard

mivanit committed:
wip jaccard

mivanit committed:
Merge branch 'main' into refactor/clustering-prereqs

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
wip jaccard

mivanit committed:
wip jaccard (plotting)

mivanit committed:
found where to increase timeout

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
wip jaccard

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
wip???

mivanit committed:
make format

mivanit committed:
fixes

mivanit committed:
typing fixes

mivanit committed:
claude doing a bunch of type hinting

mivanit committed:
Merge branch 'main' into refactor/clustering-prereqs

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
trying to get pyright passing?

this doesnt cause any issues locally :/

mivanit committed:
pyright works both locally and in CI

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
[temp] debugging disk usage

mivanit committed:
[temp] still debugging disk usage

mivanit committed:
[temp] more debug disk usage

mivanit committed:
[temp] disk debug

mivanit committed:
[temp]

mivanit committed:
[temp] remove docker? 🤔🤔🤔

mivanit committed:
[temp] uv custom stuff

mivanit committed:
ugh

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
du in root takes forever????

https://github.com/goodfire-ai/spd/actions/runs/18294192733/job/52088598056

mivanit committed:
wip

mivanit committed:
add back docker container

mivanit committed:
allow installing cpu-only torch in CI

oli-clive-griffin commented:
@mivanit can we close this for now? or do you want it for visualising the diff?
(context: I'm trying to close PRs as we have a lot open)

mivanit committed:
figure out CI disk usage by tests on main

mivanit committed:
alternate strategy for install

basically:
- torch and torchvision in a `pytorch` dep group
- `pytorch` dep group, along with `dev`, is installed by default (uv sync behavior unchanged)
- `make install-ci` recipe which first manually installs torch stuff, then uv sync but NO `pytorch` dep group

mivanit committed:
fixes to the last commit

mivanit committed:
cleanup temp changes

mivanit committed:
make in CI

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
uv sync

mivanit committed:
try to fix markupsafe?

mivanit committed:
pin markup safe with explanation

mivanit committed:
update lockfile??

mivanit committed:
Merge branch 'main' into fix/ci-disk-usage

mivanit committed:
nope i think we need the index strategy

mivanit commented:
I'd prefer it stay open for seeing the diff, but not that big of a deal either way

mivanit committed:
?

mivanit committed:
markupsafe issue

mivanit committed:
remove disk usage printing

mivanit committed:
fix pyright issue

mivanit committed:
dependency hell

mivanit committed:
fix deps???

mivanit committed:
oops, missing index strategy. moved to makefile

mivanit committed:
re-lock

mivanit committed:
make from /usr/bin/ ?

mivanit committed:
dependency hell

mivanit committed:
type checking hell

mivanit committed:
Update spd/utils/general_utils.py

Co-authored-by: Dan Braun <dan.braun@goodfire.ai>

mivanit committed:
wrap and fix Conv1D imports

mivanit committed:
minimize diff cleanup

mivanit committed:
try compile-bytecode for ci install

mivanit committed:
dont compile bytecode actually

compare:

- no compile: https://github.com/goodfire-ai/spd/actions/runs/18316145235/job/52156789078
- yes compile: https://github.com/goodfire-ai/spd/actions/runs/18316319784/job/52157420005

approximately no speedup in tests or basedpyright run, but extra 10s added to install step (where the compilation happens)

mivanit committed:
remove markupsafe constraint?

mivanit committed:
switched to use get_obj_device

mivanit committed:
remove device: torch.device type hints

see https://github.com/goodfire-ai/spd/pull/186#discussion_r2410092522

mivanit committed:
remove "distributed" test marker

mivanit committed:
fix another timeout

mivanit committed:
Merge branch 'main' into refactor/clustering-prereqs

mivanit committed:
Merge branch 'fix/ci-disk-usage' into feature/clustering

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
replace get_module_device -> get_obj_device

mivanit committed:
Merge branch 'main' into refactor/clustering-prereqs

mivanit committed:
better comments on port uniqueness

mivanit committed:
remove old markers

port uniqueness should resolve the issue, without causing slowdowns

mivanit committed:
remove timeout TODO comments

mivanit committed:
Merge branch 'refactor/clustering-prereqs' into feature/clustering

mivanit committed:
Merge branch 'main' into feature/clustering

mivanit committed:
removed checks.yaml timeout todo, clustering tests pass in ~12min

mivanit committed:
[diff-min] transformers version issue from #139 resolved

mivanit committed:
fix comment

mivanit committed:
wip jaccard

mivanit committed:
pyright fixes to jaccard, wip

mivanit committed:
Merge branch 'feature/clustering' into feature/clustering-dashboard

mivanit committed:
properly merge checks.yaml

mivanit committed:
fix torch dep issues

danbraunai-goodfire committed:
Update docs about grad syncing with DDP

mivanit committed:
wip

mivanit committed:
wip dashboard

mivanit committed:
wip

mivanit committed:
wip dashboard. bit of alpine, better filters, kinda fixed hists

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
wip

danbraunai-goodfire committed:
Mention feature/memorization-experiments in README

mivanit committed:
refactor core dashboard data gen

mivanit committed:
make format

mivanit committed:
wip big refactor of compute_max_act

mivanit committed:
more refactor, split up the big one

mivanit committed:
filtering working, table uses id & not key. more data refactor

mivanit committed:
format and bundle

mivanit committed:
wip, more spinners

mivanit committed:
wip. better logging, idk

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
make tokenizer decode go brr?

mivanit committed:
tokenizer go brr

mivanit committed:
tokenization!

mivanit committed:
wip

mivanit committed:
[!!!] cleanup

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
cleanup

mivanit committed:
wip

mivanit committed:
pyright fixes

danbraunai-goodfire committed:
Fix train and eval metrics and hidden_act_recon (#189)

danbraunai-goodfire committed:
Update canonical runs and change target model path (#197)

danbraunai-goodfire committed:
Avoid using too many processes in tests

mivanit committed:
old

mivanit committed:
Merge branch 'main' into feature/clustering

mivanit committed:
fix wandb model paths to older runs

`goodfire/spd/runs/{id}` -> `goodfire/spd-pre-Sep-2025/runs/{id}`

see https://github.com/goodfire-ai/spd/commit/5ba1d24b554ca776d94f55e4631af3ffa07e841e

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-merge-history-adapt

mivanit committed:
Merge branch 'feature/clustering-merge-history-adapt' into feature/clustering-dashboard

mivanit committed:
delete old temp json file

mivanit committed:
minimize diff

mivanit committed:
remove traces of component activation data

mivanit committed:
add `make clustering-dashboard` to CI

will need to be refactored later

mivanit committed:
comment out a lot of component-level data code

want a version that passes CI checks before a big refactor

mivanit committed:
fix test issues resulting from bytes arr -> unicode arr tok switch

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
fix makefile env usage

mivanit committed:
wandb api key for the clustering dashboard CI test

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
merge tokenization and text_processing files

mivanit committed:
fix import relating to last commit

mivanit committed:
reorg

mivanit committed:
rename frontend config file

mivanit committed:
add `DashboardConfig` instead of just tons of cli args

mivanit committed:
wip reorg

mivanit committed:
clean up interface of compute_max_activations

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
some renaming, working on using SubComponentInfo instead of str label

mivanit committed:
working on using SubComponentInfo in more places

mivanit committed:
more switching to SubComponentInfo

mivanit committed:
move `_zip_save_arr*` merge_history.py -> data_utils.py

mivanit committed:
wip

mivanit committed:
wip refactor, continued

mivanit committed:
wip

mivanit committed:
big rename `subcomponents` -> `subcomponent_keys`

in most cases, old:

`subcomponents: list[SubComponentInfo]`

now:

`subcomponent_keys: list[SubComponentKey]`

mivanit committed:
fix tests, add hashing

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
DashboardConfig inherits from BaseConfig instead of BaseModel

lets us remove the .read method

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
wip

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
allow specifying either config path or mrc cfg in pipeline cfg

mivanit committed:
[wip] reorg configs

mivanit committed:
added default `None` for slurm partition and job name prefix

mivanit committed:
refactor configs, add config tests

mivanit committed:
fix tests

mivanit committed:
allow `None` or `-1` idx_in_ensemble

- idx_in_ensemble is None iff ensemble_id is None
- idx_in_ensemble == -1 will make register_clustering_run() auto-assign next avalible index
- added tests for ensemble registry

mivanit committed:
whoops, wrong name on fixture

mivanit committed:
fix idx passed in tests when not needed

mivanit committed:
rename "mrc" -> "crc" in paths

I forgot its no longer called "MergeRunConfig"

mivanit committed:
rename merge_run_config.py -> clustering_run_config.py

mivanit committed:
fix pyright

mivanit committed:
fix idx_in_ensemble being passed in tests

mivanit committed:
rename cache dir 'merge_run_configs' -> 'clustering_run_configs'

mivanit committed:
remove component popping
changes brought in from
PR https://github.com/goodfire-ai/spd/pull/206
branch clustering/refactor-multi-batch
commit [9cbb52f](https://github.com/goodfire-ai/spd/pull/206/commits/9cbb52fd09cac8d79481a16de0a9e4c517960a33)

mivanit committed:
dont pass batch size, change not brought in here

mivanit committed:
uv sync

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
fix clustering dashboard!

mivanit committed:
add zanj.js from https://github.com/mivanit/zanj.js

mivanit committed:
zanj dep. big refactor incoming

mivanit committed:
major refactor. very broken, but not even due to refactor

something with how merge histories are saved to wandb was changed upstream. oof

mivanit committed:
continued

mivanit committed:
Merge branch 'clustering/add-mrc-inline-in-pipeline-cfg' into feature/clustering-dashboard

mivanit committed:
fix history_path extension and storage usage

mivanit committed:
dev pipeline

mivanit committed:
better config validation tests

mivanit committed:
set default base output dir

mivanit committed:
wandb use run id for clustering, TODO for spd decomp

mivanit committed:
basedpyright 1.32.0 causes issues, esp w/ wandb

https://github.com/goodfire-ai/spd/actions/runs/18719611602/job/53388090437

mivanit committed:
swap out clustering run

mivanit committed:
Merge branch 'clustering/add-mrc-inline-in-pipeline-cfg' into feature/clustering-dashboard

mivanit committed:
fix merge

mivanit committed:
fix history path

mivanit committed:
pyright fixes

mivanit committed:
pyright stuff

mivanit committed:
better asserts for debugging

mivanit committed:
[!!!] semi working again

mivanit committed:
Merge branch 'clustering/main' into clustering/add-mrc-inline-in-pipeline-cfg

mivanit committed:
wip

mivanit committed:
fix token display?

mivanit committed:
[!!!] WORKING

mivanit committed:
wip

mivanit committed:
testing via playwright

mivanit committed:
wip testing framework

mivanit committed:
wip

mivanit committed:
working but not in tests?

mivanit committed:
remove idx_in_ensemble, always auto-assigned now

see https://github.com/goodfire-ai/spd/pull/227#discussion_r2454317036

mivanit committed:
Merge branch 'clustering/main' into clustering/add-mrc-inline-in-pipeline-cfg

mivanit committed:
wip

mivanit committed:
format

mivanit committed:
something fixed???

mivanit committed:
!!!!!!!!!!!!!! PASSING?????

mivanit committed:
claude still working on stuff

mivanit committed:
make playwright tests work in CI?

mivanit committed:
fix pyright issues

mivanit committed:
only allow passing clustering run config path, not inline

see discussion at https://github.com/goodfire-ai/spd/pull/227#discussion_r2454299922

have tried to make this change as isolated as possible -- i think this was a useful feature and we may want to add it back at some point

mivanit committed:
rename run_clustering_config_path -> clustering_run_config_path

old name didnt make sense, since it should be a path to a file with a `ClusteringRunConfig`

see https://github.com/goodfire-ai/spd/pull/227#discussion_r2454299922

miv-goodfire commented:
commenting because @mivanit has reduced perms on the repo so CI won't run. hopefully interacting will enable them to run?

EDIT: commenting wont work. committing to the PR branch will though

mivanit committed:
remove junk from readme

miv-goodfire committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/add-mrc-inline-in-pipeline-cfg' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
properly install playwright

mivanit committed:
dataset streaming for dashboard tests

mivanit committed:
remove temp CI step, we now test this in pytest

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard

mivanit committed:
Merge branch 'clustering/main' into feature/clustering-dashboard
